Charged Particle Radiography: Detector Stacks

pheuer opened the pull request:
Charged particle radiography detectors are usually comprised of multiple layers of active medium, each of which detects different energies of particles based on their Bragg peak. This PR adds a couple objects that represent these detector stacks and calculates the energies deposited in each layer using stopping power data from NIST. 

This is Step 1 of adding support for detector stacks. Step 2, in a future PR, will create a new `synthetic_radiograph` function that takes a film stack object and produces a synthetic radiograph in each layer. 


~~This PR depends on #1486 and should be merged after that PR~~ Done

namurphy commented:
>     1. I need to include a couple PSTAR tables somewhere for testing and notebook documentation. Right now they are in the tests directory, but I propose that we should have a `/data` directory somewhere for this kind of lightweight data file.

Yeah, creating `plasmapy.data` to contain sample data that can be used for various analyses to try things out sounds like a really good idea to me.  We might be able to emulate [`sunpy.data`](https://github.com/sunpy/sunpy/tree/main/sunpy/data) in that regard, since they have sample observations that can be loaded with their `Map` class, for example.

>     2. I need to figure out how best to get the path to these files both in the tests and in the notebook.

A quick thought would be to look at how this is done in [`plasmapy.particles.isotopes`](https://github.com/PlasmaPy/PlasmaPy/blob/092e32bd78ed06178e801b5f4e38b5ec72bcd52a/plasmapy/particles/isotopes.py#L35) with `pkgutils.get_data` for the JSON files.   This is the sort of thing that I need to look up repeatedly, so maybe it's a good thing to put in the contributor guide.

pheuer commented:
> A quick thought would be to look at how this is done in [`plasmapy.particles.isotopes`](https://github.com/PlasmaPy/PlasmaPy/blob/092e32bd78ed06178e801b5f4e38b5ec72bcd52a/plasmapy/particles/isotopes.py#L35) with `pkgutils.get_data` for the JSON files. This is the sort of thing that I need to look up repeatedly, so maybe it's a good thing to put in the contributor guide.

Awesome, that looks easy enough! Would pkgutils be the preferred way to do it in the notebooks as well?

namurphy commented:
> Awesome, that looks easy enough! Would pkgutils be the preferred way to do it in the notebooks as well?

No idea!  😹

namurphy commented:
> > Awesome, that looks easy enough! Would pkgutils be the preferred way to do it in the notebooks as well?
> 
> No idea! joy_cat

Thinking more... or rather, thinking at all...  if we had stuff in `plasmapy.data`, then a possibility would be to do something like `from plasmapy.data import whatever`.  I'd probably emulate what SunPy does.

pheuer commented:
The SunPy download manager is quite complicated: I'll have to look into this later. It would be great if they would split that section of their code into a separate package so we could easily import it!

namurphy commented:
> The SunPy download manager is quite complicated: I'll have to look into this later. It would be great if they would split that section of their code into a separate package so we could easily import it!

I know that they have broken off some of that functionality in [parfive](https://github.com/Cadair/parfive).  Haven't used it myself, though.

pheuer commented:
@namurphy Ok, I created a new repository [PlasmaPy/sample-data](https://github.com/PlasmaPy/sample-data), a new `plasmapy/data` module, and a `data.py` file which contains a simple `get_file()` function using parfive. For now, all downloaded files are saved in `plasmapy/data/downloads/', but sunpy has a user config option to designate another folder which might be better. What do you think about this general structure?

So far fetching files this way is working for tests: I'll try it with .ipynb notebooks later.

pheuer commented:
Ok, code is working and example notebook draft is finished. [link](https://plasmapy--1274.org.readthedocs.build/en/1274/notebooks/diagnostics/charged_particle_radiography_film_stacks.html)

A few unresolved issues: 

- The "changelog: number not pull request number" error is because I branched this PR off of #1134 (because that PR changed the structure of this module), so now it has changelog entries for both. I think this should go away once #1134 is merged into main and then we merge main into this branch? 

- A few aiohttp warnings show in the notebook when downloading files that don't show when running locally. Not sure where these are coming from (the files download fine), so maybe I should just hide warnings here?

namurphy commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1115,118 +1266,375 @@ def max_deflection(self):
+        stopping_power : `~astropy.units.Quantity`
+            The stopping power in the material, multiplied by the material
+            mass density. The stopping power is tabulated in units of
+            MeV cm^2 / g, so this variable has units of MeV/cm.
```
```suggestion
            MeV cm\ :sup:`^2` / g, so this variable has units of MeV/cm.
```

namurphy commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1088,11 +1110,140 @@ def run(
+    def save_results(self, path):
+        """
+        Save the simulations results :attr:`results_dict` to a `numpy`
+        ``.npz`` file format (see `numpy.lib.format`) using `numpy.savez`.
+
+        Parameters
+        ----------
+
+        path : `str` or `os.path`
+            Either the filename (string) or an open file (file-like object)
+            where the data will be saved. If file is a string or a Path,
+            the ``.npz`` extension will be appended to the filename if
+            it is not already there.
```
```suggestion
        :file:`.npz` file format (see `numpy.lib.format`) using `numpy.savez`.

        Parameters
        ----------

        path : `str` or `os.path`
            Either the filename (string) or an open file (file-like object)
            where the data will be saved. If file is a string or a Path,
            the :file:`.npz` extension will be appended to the filename if
            it is not already there.
```

namurphy commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1115,118 +1266,375 @@ def max_deflection(self):
+            mass density. The stopping power is tabulated in units of
+            MeV cm^2 / g, so this variable has units of MeV/cm.
+
+        active : boolean, optional
+            If True, this layer is marked as an active layer. The default is True.
+
+        name : str, optional
+            An optional name for the layer.
```
```suggestion
        active : `bool`, optional
            If `True`, this layer is marked as an active layer. The default is `True`.

        name : `str`, optional
            An optional name for the layer.
```

namurphy commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1115,118 +1266,375 @@ def max_deflection(self):
+
+class Stack:
+    r"""
+    A Stack is an ordered list of Layer objects.
```
```suggestion
    A `~plasmapy.diagnostic.charged_particle_radiography.Stack` is an ordered
    list of `~plasmapy.diagnostic.charged_particle_radiography.Layer` objects.
```

namurphy commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1115,118 +1266,375 @@ def max_deflection(self):
+        Parameters
+        ----------
+
+        stack : list of Layer objects
+            This list of Layer objects defines the composition of the film stack.
```
```suggestion
        stack : list of `~plasmapy.diagnostic.charged_particle_radiography.Layer` objects
            This list of `~plasmapy.diagnostic.charged_particle_radiography.Layer`
            objects defines the composition of the film stack.
```

namurphy commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1115,118 +1266,375 @@ def max_deflection(self):
+            Energies axis over which to calculate the deposition. Units convertable
+            to eV.
+
+        return_only_active : boolean, optional
+            If True, only the deposition in layers in which the `active` property
+            is True will be returned. This is usually desirable, since particles
+            captured in other layers will not be measured. If False, deposition in
+            all layers of the stack are returned. The default is True.
```
```suggestion
        return_only_active : `bool`, optional
            If `True`, only the deposition in layers in which the `active` property
            is `True` will be returned. This is usually desirable, since particles
            captured in other layers will not be measured. If `False`, deposition in
            all layers of the stack are returned. The default is `True`.
```

namurphy replied:
```suggestion
        return_only_active : `bool`, optional
            If `True`, only the deposition in layers in which the `active` property
            is `True` will be returned. This is usually desirable, since particles
            captured in other layers will not be measured. If `False`, deposition in
            all layers of the stack are returned. The default is `True`.
```

namurphy commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1115,118 +1266,375 @@ def max_deflection(self):
+
+        Parameters
+        ----------
+        stack : list of Layer objects
+            This list of Layer objects defines the composition of the film stack.
```
```suggestion
        stack : list of `~plasmapy.diagnostic.charged_particle_radiography.Layer` objects
            This list of `~plasmapy.diagnostic.charged_particle_radiography.Layer`
            objects defines the composition of the film stack.
```

namurphy commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1115,118 +1266,375 @@ def max_deflection(self):
+            Spacing between energy bins in the calculation. Units convertable
+            to eV.
+
+        return_only_active : boolean, optional
+            If True, only the energy bands of layers in which the `active` property
+            is True will be returned. This is usually desirable, since particles
+            captured in other layers will not be measured. If False, energy bands in
+            all layers of the stack are returned. The default is True.
```
```suggestion
        return_only_active : `bool`, optional
            If `True`, only the energy bands of layers in which the `active` property
            is `True` will be returned. This is usually desirable, since particles
            captured in other layers will not be measured. If `False`, energy bands in
            all layers of the stack are returned. The default is `True`.
```

namurphy commented on changelog/1134.breaking.2.rst:
```diff
@@ -0,0 +1,2 @@
+Added the ``save_results()`` function to `~plasmapy.diagnostics.charged_particle_radiograph.Tracker` to save results to a `numpy.npz` file.
```
```suggestion
Added the `~plasmapy.diagnostics.charged_particle_radiography.Tracker.save_results` method to `~plasmapy.diagnostics.charged_particle_radiography.Tracker` to save results to a NumPy file:`.npz` file.
```

namurphy commented on changelog/1134.breaking.2.rst:
```diff
@@ -0,0 +1,2 @@
+Added the ``save_results()`` function to `~plasmapy.diagnostics.charged_particle_radiograph.Tracker` to save results to a `numpy.npz` file.
+``synthetic_radiograph`` is now a separate function in the `~plasmapy.diagnostics.charged_particle_radiograph` module that takes either a ``Tracker`` object or a result dictionary as an argument.
```
```suggestion
`~plasmapy.diagnostics.charged_particle_radiography.synthetic_radiograph` is now a separate
function in the `~plasmapy.diagnostics.charged_particle_radiography` module that takes either a
`~plasmapy.diagnostics.charged_particle_radiography.Tracker` object or a result dictionary as an
argument.
```

namurphy commented on changelog/1134.breaking.3.rst:
```diff
@@ -0,0 +1,2 @@
+`~plasmapy.diagnostics.charged_particle_radiograph.Tracker` no longer supports making changes to a `~plasmapy.diagnostics.charged_particle_radiograph.Tracker` object and re-running.
+Subsequent simulations should be performed by instantiating a new ``Tracker`` object.
```
```suggestion
Subsequent simulations should be performed by instantiating a new
`~plasmapy.diagnostics.charged_particle_radiograph.Tracker` object.
```

namurphy commented on changelog/1134.breaking.rst:
```diff
@@ -0,0 +1,2 @@
+Renamed `~plasmapy.diagnostics.proton_radiography` to `~plasmapy.diagnostics.charged_particle_radiography`,
+and renamed the ``SyntheticProtonRadiograph`` class within that module to ``Tracker``.
```
```suggestion
Renamed ``plasmapy.diagnostics.proton_radiography`` to `plasmapy.diagnostics.charged_particle_radiography`,
and renamed the ``SyntheticProtonRadiograph`` class within that module to `~plasmapy.diagnostics.charged_particle_radiograph.Tracker`.
```

namurphy commented on changelog/1274.feature.rst:
```diff
@@ -0,0 +1,2 @@
+Added the ``Stack`` and ``Layer` objects to the `~plasmapy.diagnostics.charged_particle_radiography` module, which represent a stack of detector media layers.
+The ``deposition_curve`` and ``energy_bands`` methods of ``Stack``  calculate the particle energies deposited in each detector layer.
```
```suggestion
Added the `~plasmapy.diagnostics.charged_particle_radiograph.Stack` and 
`~plasmapy.diagnostics.charged_particle_radiograph.Layer` objects to the
`~plasmapy.diagnostics.charged_particle_radiography` module, which 
represent a stack of detector media layers. The 
`~plasmapy.diagnostics.charged_particle_radiography.Stack.deposition_curve` and 
`~plasmapy.diagnostics.charged_particle_radiography.Stack.energy_bands` methods of 
`~plasmapy.diagnostics.charged_particle_radiography.Stack` calculate the 
particle energies deposited in each detector layer.
```

namurphy commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1088,11 +1110,140 @@ def run(
+    def save_results(self, path):
+        """
+        Save the simulations results :attr:`results_dict` to a `numpy`
+        ``.npz`` file format (see `numpy.lib.format`) using `numpy.savez`.
```
```suggestion
        :file:`.npz` file format (see `numpy.lib.format`) using `numpy.savez`.
```

namurphy commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1088,11 +1110,140 @@ def run(
+        path : `str` or `os.path`
+            Either the filename (string) or an open file (file-like object)
+            where the data will be saved. If file is a string or a Path,
+            the ``.npz`` extension will be appended to the filename if
```
```suggestion
            the :file:`.npz` extension will be appended to the filename if
```

namurphy commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1088,11 +1110,140 @@ def run(
+
+        path : `str` or `os.path`
+            Either the filename (string) or an open file (file-like object)
+            where the data will be saved. If file is a string or a Path,
```
```suggestion
            where the data will be saved. If file is a string or a path,
```

namurphy commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1115,118 +1266,375 @@ def max_deflection(self):
+        detector plane and :math:`I_0` is the intensity on the detector
+        plane in the absence of simulated fields. Default is `False`.
+
+        If the `Intensity` histogram contains zeros, `OD` will contain
+        -`~numpy.inf` values. These can be easily replaced with zeros
+        if desired for plotting using `~numpy.nan_to_num`
+
+        ``OD = np.nan_to_num(OD, neginf=0)``
```
```suggestion
        If the ``Intensity`` histogram contains zeros, ``OD`` will contain
        -`~numpy.inf` values. These can be easily replaced with zeros
        if desired for plotting using `~numpy.nan_to_num`

        .. code-block::
           OD = np.nan_to_num(OD, neginf=0)
```

namurphy commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1115,118 +1266,375 @@ def max_deflection(self):
+            MeV cm^2 / g, so this variable has units of MeV/cm.
+
+        active : boolean, optional
+            If True, this layer is marked as an active layer. The default is True.
```
```suggestion
            If `True`, this layer is marked as an active layer. The default is `True`.
```

namurphy commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1115,118 +1266,375 @@ def max_deflection(self):
+        active : boolean, optional
+            If True, this layer is marked as an active layer. The default is True.
+
+        name : str, optional
```
```suggestion
        name : `str`, optional
```

namurphy reviewed (commented):
Thank you for working on this...I'm really looking forward to it!  

I had a chance to go through the docstrings and make some suggested changes related to reST formatting (mostly described in the very recently updated [doc guide](https://docs.plasmapy.org/en/latest/development/doc_guide.html)).  All of the suggestions are minor or trivial so it shouldn't take long to go through them. 

Thank you again!

pheuer commented:
Thanks @namurphy ! 

At our diagnostics meeting this week, @rocco8773 suggested that the `get_file` function and downloaded files might more properly belong in the `util` subpackage (at least while they remain primarily non user facing). This avoids adding to the PLEP. Does that make sense to you?

pheuer commented:
Also, does anyone know why this code-block is raising an error? It looks ok to me...

```
/home/runner/work/PlasmaPy/PlasmaPy/plasmapy/diagnostics/charged_particle_radiography.py:docstring of plasmapy.diagnostics.charged_particle_radiography.synthetic_radiograph:34: WARNING: Error in "code-block" directive:
maximum 1 argument(s) allowed, 4 supplied.

.. code-block::
   OD = np.nan_to_num(OD, neginf=0)
Loading template...
Finding news fragments...
ERROR: InvocationError for command /home/runner/work/PlasmaPy/PlasmaPy/.tox/build_docs/bin/sphinx-build docs docs/_build/html -W --keep-going -b html -q (exited with code 1)
```

namurphy commented:
> Also, does anyone know why this code-block is raising an error? It looks ok to me...

Looks like you got it!  There needs to be a blank line between the directive and the content.

namurphy commented:
> At our diagnostics meeting this week, @rocco8773 suggested that the `get_file` function and downloaded files might more properly belong in the `util` subpackage (at least while they remain primarily non user facing). This avoids adding to the PLEP. Does that make sense to you?

That does sound reasonable to me.  We'll probably need a `plasmapy.io` subpackage eventually, but for the moment this would be a good place to put it.

pheuer commented:
This linter is in error: the line in question is not an fstring but a template passed to `tqdm`

```
linters run-test: commands[1] | flake8 /home/runner/work/PlasmaPy/PlasmaPy/plasmapy --count --show-source --statistics
plasmapy/diagnostics/charged_particle_radiography.py:1075:24: FS003 f-string missing prefix
            bar_format="{l_bar}{bar}{n:.1e}/{total:.1e} {unit}",
                       ^
1     FS003 f-string missing prefix
```

namurphy commented:
> This linter is in error: the line in question is not an fstring but a template passed to `tqdm`
> 
> ```
> linters run-test: commands[1] | flake8 /home/runner/work/PlasmaPy/PlasmaPy/plasmapy --count --show-source --statistics
> plasmapy/diagnostics/charged_particle_radiography.py:1075:24: FS003 f-string missing prefix
>             bar_format="{l_bar}{bar}{n:.1e}/{total:.1e} {unit}",
>                        ^
> 1     FS003 f-string missing prefix
> ```

Sorry about that!  I forgot to mention that after #1281 was merged we'd need to change [this line of  `setup.cfg`](https://github.com/PlasmaPy/PlasmaPy/blob/727513de1af001b618a5cfbe8a787bcc313a8351/setup.cfg#L206) to reflect the change in the filename, after doing a `git pull plasmapy main`.

I'm in the process of adding some new static analysis tools, so if you have other problems with linters in the next few days, feel free to ping me about it since I've been trying to avoid the files you've been working on.

pheuer commented:
>  so we have to tell it to ignore that particular check on this file.

Got it - thanks!

pheuer commented:
The data downloading portion of this branch is unfortunately broken. There are two issues:

- Github does something weird with file headers. The files uploaded are not compressed and do not have a Content-Type 'gzip' header, but if I download them from the following link they do have that header and that breaks parfive: 
"https://github.com/PlasmaPy/sample-data/raw/main/data/"

I'm using this other service link as a workaround, but that isn't ideal

"https://gitcdn.link/repo/PlasmaPy/data/main/data/"

- The current program doesn't thrown an error if you have it download a non-existent file. I'm not sure what is going on here. 

I need to look more into this and see if there is a better way to do this.

Paging @namurphy for ideas

pheuer commented:
I've split the file handling out from this PR into #1486 , so that PR should be merged before this one.

pheuer commented:
@rocco8773 @namurphy This PR has also been waiting for review for a long time - can we try to get this merged in the next month? I've just updated to main, so it should be ready to go.

namurphy commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1419,3 +1422,249 @@ def synthetic_radiograph(
+    r"""
+    An ordered list of
+    `~plasmapy.diagnostics.charged_particle_radiography.Layer` objects.
+
```
```suggestion
```

StanczakDominik commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1419,3 +1422,249 @@ def synthetic_radiograph(
+        A list of the `~plasmapy.diagnostics.charged_particle_radiography.Layer` objects that make up the film stack.
+    """
+
+    def __init__(self, layers):
```
```suggestion
    def __init__(self, layers: List[Layer]):
```
(would require `from typing import List`)

StanczakDominik commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1419,3 +1422,249 @@ def synthetic_radiograph(
+        return len(self._layers)
+
+    @property
+    def nactive(self):
```
Maybe `num_active`?

pheuer replied:
Done

StanczakDominik commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1419,3 +1422,249 @@ def synthetic_radiograph(
+        r"""
+        The total thickness of the stack.
+        """
+        thickness = np.array([layer.thickness.to(u.mm).value for layer in self._layers])
+        return np.sum(thickness) * u.mm
```
We can probably skip the element-wise unit operations and operate on the whole array:
```suggestion
        thickness = u.Quantity([layer.thickness for layer in self._layers])
        return thickness.sum().to(u.mm)
```

StanczakDominik commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1419,3 +1422,249 @@ def synthetic_radiograph(
+        Parameters
+        ----------
+
+        stack : list of `~plasmapy.diagnostics.charged_particle_radiography.Layer` objects
+            This list of `~plasmapy.diagnostics.charged_particle_radiography.Layer`
+            objects defines the composition of the film stack.
+
```
```suggestion
```
This seems to be encompassed by `self` nowadays :)

StanczakDominik commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1419,3 +1422,249 @@ def synthetic_radiograph(
+        -------
+
+        deposited : `~numpy.ndarray`, shape [nlayers, nenergies]
+            The fraction of an ensemble of each energy that will be deposited in
```
Fraction of an ensemble of each energy? I'm just going by the docs for now, and it's not quite obvious to me what this means.

pheuer replied:
Agreed - so what I'm trying to say is that if you start with an ensemble of N particles of energy E, this will calculate the fraction of N stopped in each layer. Then it does that for all the energies E to get a 2D array of 'deposition curves'.

I'll take a stab at rephrasing to that efffect.

StanczakDominik commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1419,3 +1422,249 @@ def synthetic_radiograph(
+
+            # Slice the layer into sublayer dx thick
+            nsublayers = int(
+                np.floor(layer.thickness.to(u.um).value / dx.to(u.um).value)
```
```suggestion
                (layer.thickness // dx).si
```

StanczakDominik commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1419,3 +1422,249 @@ def synthetic_radiograph(
+        # Deposited energy in MeV
+        deposited = np.zeros([len(self._layers), energies.size])
+
+        for i, layer in enumerate(self._layers):
```
I think I have an idea for how to optimize this. I'll come back to it tomorrow.

StanczakDominik commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1419,3 +1422,249 @@ def synthetic_radiograph(
+        # Normalize the deposited energy array so that each number represents
+        # the fraction of a population of particles of that energy stopped
+        # in that layer.
+        deposited = deposited / np.sum(deposited, axis=0)
```
A tiny bit less code 😜 
```suggestion
        deposited /= np.sum(deposited, axis=0)
```

StanczakDominik commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1419,3 +1422,249 @@ def synthetic_radiograph(
+
+        Parameters
+        ----------
+        stack : list of `~plasmapy.diagnostics.charged_particle_radiography.Layer` objects
+            This list of `~plasmapy.diagnostics.charged_particle_radiography.Layer`
+            objects defines the composition of the film stack.
+
```
```suggestion
        stack : list of `~plasmapy.diagnostics.charged_particle_radiography.Layer` objects
            This list of `~plasmapy.diagnostics.charged_particle_radiography.Layer`
            objects defines the composition of the film stack.

```

StanczakDominik commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1419,3 +1422,249 @@ def synthetic_radiograph(
+
+        return deposited
+
+    def energy_bands(self, energy_range, dE, dx=1 * u.um, return_only_active=True):
```
As before, annotations would be nice to have here!

StanczakDominik commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1419,3 +1422,249 @@ def synthetic_radiograph(
+
+        energies = (
+            np.arange(
+                energy_range[0].to(u.MeV).value,
+                energy_range[1].to(u.MeV).value,
```
This will do:
```suggestion
                *energy_range.to(u.MeV).value,
```

StanczakDominik reviewed (changes requested):
Mostly done; I didn't manage to read through the notebook yet. Got a bunch of suggestions, mostly typo fixes, minor improvements, that kind of thing. We should be able to get it polished up and merged quickly.

StanczakDominik commented:
BTW, sorry, I was reviewing this late tomorrow and I was kinda grumpy. Would you like me to apply the fixes above myself?

pheuer commented on plasmapy/diagnostics/charged_particle_radiography.py:
```diff
@@ -1419,3 +1422,249 @@ def synthetic_radiograph(
     return h * u.m, v * u.m, intensity
+
+
+# *************************************************************************
+# Film Stack & Layers
+# *************************************************************************
```
Yeah, I'm good with that!

pheuer replied:
Done. I didn't make it private because this isn't actually imported by the charged_particle_radiography file (yet) but is a separate and related tool.

pheuer replied:
So, I moved this over but since `_detector_stacks` is private I would still like users to import from the `charged_particle_radiography` module, and also for references like 
`~plasmapy.diagnostics.charged_particle_radiography.Layer` to work. What's the best way to do that?

pheuer commented on plasmapy/diagnostics/tests/test_charged_particle_radiography.py:
```diff
@@ -133,7 +135,7 @@ def _test_grid(
     """
     Run a simulation through an example with parameters optimized to
-    sum up to a lineout along x. The goal is to run a relatively fast
+    sum up to a lineout along x. The goal is to run a realtively fast
```
Never underestimate how bad I am at spelling!

pheuer commented on plasmapy/diagnostics/tests/test_charged_particle_radiography.py:
```diff
@@ -579,31 +579,22 @@ def test_optical_density_histogram(self):
-        # returned from the function
+        intensity = intensity_results[2]
+        zero_mask = intensity == 0
+        i0 = np.mean(intensity[~zero_mask])
+        od = -np.log10(intensity / i0)
```
How about `Initial_intensity` and `optical_density`?

pheuer commented:
> BTW, sorry, I was reviewing this late tomorrow and I was kinda grumpy. Would you like me to apply the fixes above myself?

No worries - I can apply them! Thanks for the review :)

StanczakDominik commented:
The doc build warnings are most likely going to be related to the notes in https://github.com/PlasmaPy/PlasmaPy/commit/4e3d52b76fb976dc7ab40cb9248a056edc36360e ...

pheuer commented:
> The doc build warnings are most likely going to be related to the notes in [4e3d52b](https://github.com/PlasmaPy/PlasmaPy/commit/4e3d52b76fb976dc7ab40cb9248a056edc36360e) ...

Yes, in this case I think my confusion is that the actual object isn't in `charged_particle_radiography` anymore (the code is in `_detector_stacks`). So, in sphinx to I have to explicitly reference the object in `_detector_stacks` or is there a way I can make it appear to be in `charged_particle_radiography`?

StanczakDominik commented:
I'll figure that out later today :)

StanczakDominik commented:
So the main difference now is that we're checking the SCM version CONDITIONALLY on being in an editable install, because of the manifest.in file

rocco8773 commented on plasmapy/diagnostics/_detector_stacks.py:
```diff
@@ -0,0 +1,257 @@
+        energy_axis : `~astropy.units.Quantity`
+            The energies corresponding to the stopping power array.
+
+        stopping_power : `~astropy.units.Quantity`
+            The stopping power in the material, multiplied by the material
+            mass density. The stopping power is tabulated in units of
+            MeV cm\ :sup:`^2` / g, so this variable has units of MeV/cm.
```
This is tripping me up a bit since the term "stopping power" is being used for two different concepts, linear and mass stopping power, without clarification.  I think there's several options to proceed with, but I'm not sure with is the best option at the moment...

1. Keep the keyword argument `stopping_power`, but update the description to enforce that you mean the linear stopping power.
2. Keep the keyword argument `stopping_power`, but allow it to be either the linear or the mass stopping power.  This would also require an additional argument for the mass density of the film (It could possibly be optional 🤔).  However, the class attributes would need to be stored as `linear_stopping_power` and `mass_stopping_power`, instead of just `stopping_power`.

---

After writing this this, I think I'm favoring option 2.  Thoughts?

StanczakDominik commented:
By the way, I went through the notebook and it looks great! Still trying to figure out the doc warnings. @rocco8773, could you take a look?

rocco8773 commented:
The doc warnings are coming from how the functionality is linked versus where it's actually defined. At APS, @pheuer and I talked about moving all this functionality into its own subpackage. Assuming that is done correctly, all the doc warnings will fix themselves.

namurphy commented:
@pheuer — We were thinking of doing the 0.9.0 release fairly shortly, likely in the next few days. Did you want to try to get this in before then, or wait until 0.10.0?  We're probably going to try releases every 3-4 months, so it wouldn't be that long of a wait.

StanczakDominik commented:
> every 3-4 months

I'd honestly like to try a bit more often, maybe ~2 months - might motivate me to crunch down on those features more often! :)

pheuer commented:
@namurphy I probably won't be able to finish this in the next few days - it's ok with me if it waits for 0.10.

StanczakDominik commented:
The fact that we're seeing the old azure test runner here is... well, extremely weird to say the least. I'll push a merge of main into this branch.

pheuer commented:
@rocco8773 I added support for the `stopping_power` to be either linear or mass, with an optional mass_density keyword that is required if the mass stopping power is provided (checked by units). I also converted everything internally to SI units. 

I split charged particle radiography off into its own module as we discussed at APS. I like this better. One question - what should we rename the file with `Tracker` and `synthetic_radiograph` in it? Right now I've renamed it `synthetic_radiography`, but open to suggestions.

pheuer commented:
@rocco8773 Any updates on this PR?

pheuer commented:
@namurphy and @StanczakDominik , do either of you have a preference regarding this naming question? 

> I split charged particle radiography off into its own module as we discussed at APS. I like this better. One question - what should we rename the file with `Tracker` and `synthetic_radiograph` in it? Right now I've renamed it `synthetic_radiography`, but open to suggestions.

Otherwise I think this PR should be nearly finished

namurphy commented:
> @namurphy and @StanczakDominik , do either of you have a preference regarding this naming question?
> 
> > I split charged particle radiography off into its own module as we discussed at APS. I like this better. One question - what should we rename the file with `Tracker` and `synthetic_radiograph` in it? Right now I've renamed it `synthetic_radiography`, but open to suggestions.

Hm...I can't say I know the subject matter well enough to have much of an opinion.  Having the file itself be named `synthetic_radiography` seems quite reasonable to me.

> Otherwise I think this PR should be nearly finished

Awesome!

pheuer commented:
> Hm...I can't say I know the subject matter well enough to have much of an opinion. Having the file itself be named `synthetic_radiography` seems quite reasonable to me.

Great, I'll move forward with that, get all the broken links resolved and then hopefully we can get it merged.

namurphy commented on plasmapy/diagnostics/charged_particle_radiography/detector_stacks.py:
```diff
@@ -0,0 +1,292 @@
+
+        Returns
+        -------
+
```
```suggestion
```

namurphy commented on plasmapy/diagnostics/charged_particle_radiography/detector_stacks.py:
```diff
@@ -0,0 +1,292 @@
+
+        Parameters
+        ----------
+
```
```suggestion
```

namurphy commented on plasmapy/diagnostics/charged_particle_radiography/tests/test_detector_stacks.py:
```diff
@@ -0,0 +1,124 @@
+    # Test deposition curves
+    deposition_curves = hdv2_stack.deposition_curves(energies, return_only_active=False)
+
+    # Test that integral over all layers for each particle species is unity
+    integral = np.sum(deposition_curves, axis=0)
+    assert np.allclose(integral, 1.0)
```
```suggestion
    integral = np.sum(deposition_curves, axis=0)
    assert np.allclose(integral, 1.0), (
        "The integral over all layers for each particle species "
        "is not unity."
    )
```

namurphy commented on plasmapy/diagnostics/charged_particle_radiography/tests/test_synthetic_radiography.py:
```diff
@@ -624,7 +614,7 @@ def test_saving_output(tmp_path):
 
     # Save result
-    path = str(tmp_path / "temp.npz")
+    path = os.path.join(tmp_path, "temp.npz")
```
```suggestion
    path = tmp_path / "temp.npz"
```
The `tmp_path` fixture returns a `pathlib.Path` object, so it should be possible to do this instead of needing `os.path`.

namurphy commented on plasmapy/diagnostics/charged_particle_radiography/detector_stacks.py:
```diff
@@ -0,0 +1,292 @@
+
+        Returns
+        -------
+
```
```suggestion
```

namurphy commented on plasmapy/diagnostics/charged_particle_radiography/tests/test_detector_stacks.py:
```diff
@@ -0,0 +1,124 @@
+"""
+Tests for proton radiography functions
```
Not for the first time, I am lamenting the lack of emoji related to charged particle radiography!  😿

pheuer replied:
:)

namurphy reviewed (commented):
Overall this is looking pretty close!  This is a fairly superficial review since I'm not too familiar with the subject matter.  All of my suggestions are pretty minor, and some of the docstring changes are associated with #1757 which is using some conventions from matplotlib.

Another thing I'm wondering (unrelated to this pull request) is if it would be good to add "mass stopping power" as a physical type to [`astropy.units.physical`](https://github.com/astropy/astropy/blob/main/astropy/units/physical.py).  I imagine mass stopping power would be more useful to have it in there than "absement"!

And thank you again for doing this!

pheuer commented on docs/ad/diagnostics/charged_particle_radiography/synthetic_radiography.rst:
```diff
@@ -1,16 +1,17 @@
 
 *************************************
-Charged particle radiography analysis
+Synthetic Charged Particle Radiography
```
That's the idea!

pheuer commented on plasmapy/diagnostics/charged_particle_radiography/detector_stacks.py:
```diff
@@ -0,0 +1,292 @@
+    ----------
+
+    layers : list of `~plasmapy.diagnostics.charged_particle_radiography.Layer` objects
+        A list of the `~plasmapy.diagnostics.charged_particle_radiography.Layer` objects that make up the film stack.
```
My least favorite PEP

pheuer commented on plasmapy/diagnostics/tests/test_charged_particle_radiography.py:
```diff
@@ -869,3 +860,98 @@ def test_add_wire_mesh():
+    test_gaussian_sphere_analytical_comparison()
+    test_cannot_modify_simulation_after_running()
+    """
+    pass
```
Yeah, the only reason I do this during development is because it's easier to quickly add debugging lines. But I should remove them before the PR is put out.

pheuer commented on plasmapy/diagnostics/tests/test_charged_particle_radiography.py:
```diff
@@ -579,31 +579,22 @@ def test_optical_density_histogram(self):
+        intensity = intensity_results[2]
+        zero_mask = intensity == 0
+        i0 = np.mean(intensity[~zero_mask])
+        od = -np.log10(intensity / i0)
```
Something weird happened where a lot of your comments seem to be on a several months old version of this file? Anyway, we changed this variable name to `optical_density`

pheuer commented on plasmapy/diagnostics/tests/test_charged_particle_radiography.py:
```diff
@@ -869,3 +860,98 @@ def test_add_wire_mesh():
+    assert hdv2_stack.nactive == 10
+
+    # Test thickness property
+    assert np.isclose(hdv2_stack.thickness.to(u.mm).value, 1.19)
```
Done

pheuer commented:
@namurphy Do you know how I can tell the linter to ignore this line? This isn't an fstring, it's a format string for the tqdm progress bar (the variables aren't defined in my code, so adding the  'f' prefix raises exceptions)

```
plasmapy/diagnostics/charged_particle_radiography/synthetic_radiography.py:1067:24: FS003 f-string missing prefix
            bar_format="{l_bar}{bar}{n:.1e}/{total:.1e} {unit}",
                       ^
1     FS003 f-string missing prefix
```

namurphy commented on plasmapy/diagnostics/charged_particle_radiography/synthetic_radiography.py:
```diff
@@ -1066,7 +1060,7 @@ def run(
             desc="Particles on grid",
             unit="particles",
-            bar_format="{l_bar}{bar}{n:.1e}/{total:.1e} {unit}",
+            bar_format="{l_bar}{bar}{n:.1e}/{total:.1e} {unit}",  # noqa
```
```suggestion
            bar_format="{l_bar}{bar}{n:.1e}/{total:.1e} {unit}",  # noqa: FS003
```
If we add the error code, it will still do the quality assurance checks for everything else.

pheuer replied:
Oh nice!

namurphy commented:
The Python 3.10 error appears to be related to #750 and unrelated to this PR.  I'll look into it later.

StanczakDominik commented:
I merged `main` into this to hopefully solve the CI errors.

namurphy commented on docs/ad/diagnostics/charged_particle_radiography/synthetic_radiography.rst:
```diff
@@ -1,12 +1,12 @@
-.. _charged_particle_radiography:
+.. _synthetic_radiography:
 
 *************************************
```
```suggestion
**************************************
```

StanczakDominik replied:
This right here is what I absolutely hate about restructuredtext.

StanczakDominik replied:
I'm just going to add myst-parser to our doc dependencies so that we may also write these in Markdown. Hold my green tea.

namurphy commented on docs/ad/diagnostics/charged_particle_radiography/synthetic_radiography.rst:
```diff
@@ -1,12 +1,12 @@
 *************************************
-Charged particle radiography analysis
+Synthetic Charged Particle Radiography
 *************************************
 
```
```suggestion
**************************************
```

namurphy commented on plasmapy/diagnostics/charged_particle_radiography/detector_stacks.py:
```diff
@@ -0,0 +1,286 @@
+        stopping_power : `~astropy.units.Quantity`
+            The stopping power in the material. Either the linear stopping
+            power (units of J/m) or the mass stopping power
+            (units convertible to J m\ :sup:`^2` / kg) can be provided. If the
```
```suggestion
            (units convertible to J m\ :sup:`2` / kg) can be provided. If the
```

namurphy commented on plasmapy/diagnostics/charged_particle_radiography/detector_stacks.py:
```diff
@@ -0,0 +1,286 @@
+
+class Stack:
+    r"""
+    An ordered list of
+    `~plasmapy.diagnostics.charged_particle_radiography.detector_stacks.Layer` objects.
+
+    Parameters
+    ----------
+    layers : list of `~plasmapy.diagnostics.charged_particle_radiography.detector_stacks.Layer` objects
+        A list of the `~plasmapy.diagnostics.charged_particle_radiography.detector_stacks.Layer`
+        objects that make up the film stack.
```
```suggestion
    An ordered list of |Layer| objects.

    Parameters
    ----------
    layers : list of |Layer|
        The objects that make up the film stack.

    .. |Layer| replace:: `~plasmapy.diagnostics.charged_particle_radiography.detector_stacks.Layer`
```
[optional] Since the full namespace is really long, I was wondering if a local reST substition would make the bulk of the docstring more readable (and also fit within ≲72 characters).  This suggestion would also keep type information in the type specification, and the meaning in the parameter description.

pheuer replied:
Yeah good idea, I've now done this for a few other points too

namurphy commented on plasmapy/diagnostics/charged_particle_radiography/synthetic_radiography.py:
```diff
@@ -90,8 +89,9 @@ class Tracker:
         based on its units. Valid geometries are:
 
+        * Cartesian (x,y,z) : (meters, meters, meters)
         * Cartesiaical (r, theta, z) : (meters, radians, meters)
```
```suggestion
        * Cylindrical (r, theta, z) : (meters, radians, meters)
```
Is this supposed to be "cylindrical" here?

pheuer replied:
>_>

pheuer commented:
Something appears to not quite be working with the substutition for Layer? 

![image](https://user-images.githubusercontent.com/32618747/207213358-e519be8c-dec5-4445-b605-452ed3683497.png)

pheuer commented:
I think I fixed the issue with the substitution by moving it earlier before the parameters block.

Here are the remaining doc errors, broadly in two categories

A number of this first set relate to things like ``nenergies`` being marked like markdown but not corresponding to any actual variable?

```
/home/runner/work/PlasmaPy/PlasmaPy/plasmapy/diagnostics/charged_particle_radiography/detector_stacks.py:docstring of plasmapy.diagnostics.charged_particle_radiography.detector_stacks.Stack.deposition_curves:2: WARNING: py:obj reference target not found: nenergies
/home/runner/work/PlasmaPy/PlasmaPy/plasmapy/diagnostics/charged_particle_radiography/detector_stacks.py:docstring of plasmapy.diagnostics.charged_particle_radiography.detector_stacks.Stack.deposition_curves:17: WARNING: py:obj reference target not found: nlayers
/home/runner/work/PlasmaPy/PlasmaPy/plasmapy/diagnostics/charged_particle_radiography/detector_stacks.py:docstring of plasmapy.diagnostics.charged_particle_radiography.detector_stacks.Stack.deposition_curves:17: WARNING: py:obj reference target not found: nenergies
/home/runner/work/PlasmaPy/PlasmaPy/plasmapy/diagnostics/charged_particle_radiography/detector_stacks.py:docstring of plasmapy.diagnostics.charged_particle_radiography.detector_stacks.Stack.energy_bands:17: WARNING: py:obj reference target not found: nlayers

```


The rest are related to broken links in the What's New sections. I would suggest just removing this folder from the space that CI checks? Otherwise links will of course break every time we refactor anything. 


```
/home/runner/work/PlasmaPy/PlasmaPy/docs/whatsnew/0.7.0.rst:42: WARNING: py:func reference target not found: plasmapy.diagnostics.charged_particle_radiography.synthetic_radiograph
/home/runner/work/PlasmaPy/PlasmaPy/docs/whatsnew/0.7.0.rst:42: WARNING: py:obj reference target not found: plasmapy.diagnostics.charged_particle_radiography.Tracker
/home/runner/work/PlasmaPy/PlasmaPy/docs/whatsnew/0.7.0.rst:42: WARNING: py:obj reference target not found: plasmapy.diagnostics.charged_particle_radiography.Tracker.results_dict
/home/runner/work/PlasmaPy/PlasmaPy/docs/whatsnew/0.7.0.rst:49: WARNING: py:obj reference target not found: plasmapy.diagnostics.charged_particle_radiography.Tracker
/home/runner/work/PlasmaPy/PlasmaPy/docs/whatsnew/0.7.0.rst:53: WARNING: py:obj reference target not found: plasmapy.diagnostics.charged_particle_radiography.Tracker
/home/runner/work/PlasmaPy/PlasmaPy/docs/whatsnew/0.7.0.rst:53: WARNING: py:obj reference target not found: plasmapy.diagnostics.charged_particle_radiography.Tracker
/home/runner/work/PlasmaPy/PlasmaPy/docs/whatsnew/0.7.0.rst:116: WARNING: py:meth reference target not found: plasmapy.diagnostics.charged_particle_radiography.Tracker.save_results
/home/runner/work/PlasmaPy/PlasmaPy/docs/whatsnew/0.7.0.rst:116: WARNING: py:obj reference target not found: plasmapy.diagnostics.charged_particle_radiography.Tracker
/home/runner/work/PlasmaPy/PlasmaPy/docs/whatsnew/0.7.0.rst:153: WARNING: py:obj reference target not found: plasmapy.diagnostics.charged_particle_radiography.synthetic_radiograph
/home/runner/work/PlasmaPy/PlasmaPy/docs/whatsnew/0.7.0.rst:[28](https://github.com/PlasmaPy/PlasmaPy/actions/runs/3682075943/jobs/6229365410#step:6:29)9: WARNING: py:obj reference target not found: plasmapy.diagnostics.charged_particle_radiography.synthetic_radiograph

```

StanczakDominik commented:
Wow, pretty obvious in retrospect! The thing is, this is the actual sphinx build we're checking for the warnings. Obviously it's not enough anymore. I guess instead of running `sphinx-build -W` we could do `sphinx-build -w output.log`, then grep (well, regex) for `PlasmaPy/docs/whatsnew/.*\.rst:.* WARNING: py:obj reference target not found .*` and raise if any are found. Can't be that hard.

pheuer commented:
> On the `nenergies` stuff... @namurphy, @rocco8773?

These lines look like this: 

```
energies : (`nenergies`,) `~astropy.units.Quantity` array
            Energies axis over which to calculate the deposition. Units
            convertible to J.
```

There is no variable nenergies. I could just remove the `` which would be the easiest solution?

StanczakDominik commented:
Let's try it!

StanczakDominik commented:
Actually, wisdom by @namurphy: do ` ``nenergies``, ` and it should work!

namurphy commented on plasmapy/diagnostics/charged_particle_radiography/detector_stacks.py:
```diff
@@ -0,0 +1,287 @@
+    r"""
+    An ordered list of |Layer| objects.
+
+    .. |Layer| replace:: `~plasmapy.diagnostics.charged_particle_radiography.detector_stacks.Layer`
```
It looks like this substitution is getting used in the narrative documentation as well.  To use this more broadly, we can define it in `docs/common_links.py`.

pheuer replied:
Done - I didn't see a section for the diagnostics sub package so I added one

namurphy replied:
Awesome!

namurphy commented:
Still getting a doc build error, seemingly due to a reST formatting problem.  
```
WARNING: :40: (WARNING/2) Bullet list ends without a blank line; unexpected unindent.
WARNING: :47: (WARNING/2) Definition list ends without a blank line; unexpected unindent.
```
It would have been helpful if it had said the file that the issue is occurring in...sigh!

namurphy commented on changelog/1274.breaking.rst:
```diff
@@ -0,0 +1,6 @@
+Moved the charged particle radiography analysis codes into a new module
+`~plasmapy.diagnostics.charged_particle_radiography` containing synthetic
+radiograpy tools in
+`~plasmapy.diagnostics.charged_particle_radiography.synthetic_radiography`
+and detector stack calculation tools in
+`~plasmapy.diagnostics.charged_particle_radiography.detector_stacks`
```
```suggestion
Moved the charged particle radiography analysis codes into a new module `~plasmapy.diagnostics.charged_particle_radiography` containing synthetic radiography tools in `~plasmapy.diagnostics.charged_particle_radiography.synthetic_radiography` and detector stack calculation tools in `~plasmapy.diagnostics.charged_particle_radiography.detector_stacks`.
```
One *possible* cause of the doc failure is that it is interpreting this as a list or something.  Maybe.  In the past I think I was able to fix something like this by putting the changelog entry all on one line.  Not sure why.  But, could be worth trying out.

namurphy commented on changelog/1274.feature.rst:
```diff
@@ -0,0 +1,10 @@
+Added the `~plasmapy.diagnostics.charged_particle_radiography.detector_stacks.Stack` and
+|Layer| objects to the
+`~plasmapy.diagnostics.charged_particle_radiography` module, which
+represent a stack of detector media layers. The
+`~plasmapy.diagnostics.charged_particle_radiography.detector_stacks.Stack.deposition_curves`
+and
+`~plasmapy.diagnostics.charged_particle_radiography.detector_stacks.Stack.energy_bands`
+methods of
+`~plasmapy.diagnostics.charged_particle_radiography.detector_stacks.Stack`
+calculate the particle energies deposited in each detector layer.
```
```suggestion
Added the `~plasmapy.diagnostics.charged_particle_radiography.detector_stacks.Stack` and
|Layer| objects to the `~plasmapy.diagnostics.charged_particle_radiography` module, which represent a stack of detector media layers. The `~plasmapy.diagnostics.charged_particle_radiography.detector_stacks.Stack.deposition_curves` and `~plasmapy.diagnostics.charged_particle_radiography.detector_stacks.Stack.energy_bands`
methods of `~plasmapy.diagnostics.charged_particle_radiography.detector_stacks.Stack` calculate the particle energies deposited in each detector layer.
```

pheuer commented:
> It would have been helpful if it had said the file that the issue is occurring in...sigh!

Indeed...let's see if this works!

pheuer commented:
@namurphy that worked! Merging before the linters change their mind :)

namurphy commented:
> @namurphy that worked! Merging before the linters change their mind :)

Congratulations on getting it merged!
